07_errors/willshen8/syncStore.go

```diff
@@ -0,0 +1,79 @@
+package main
+
+import (
+	"strconv"
+	"sync"
+)
+
+type SyncStore struct {
+	syncStore sync.Map
+	nextID    uint32
+	mu        sync.Mutex
+}
+
+// NewMapStore initialise a new SyncStore
+func NewSyncStore() *SyncStore {
+	return &SyncStore{syncStore: sync.Map{}}
+}
+
+// CreatePuppy create a new puppy and store in mapStore.
+func (s *SyncStore) CreatePuppy(p *Puppy) (uint32, error) {
+	if i, err := strconv.Atoi(p.Value); err == nil {
+		if i < 0 {
+			return 0, &Error{
+				Message: "Puppy value can't be less than 0.",
+				Code:    1,
+			}
+		}
+	}
+	s.mu.Lock()
+	s.nextID++
+	p.ID = s.nextID
+	s.syncStore.Store(p.ID, *p)
+	s.mu.Unlock()
+	return p.ID, nil
+}
+
+// ReadPuppy read a puppy given its id.
+// It returns the pointer to that puppy.
+func (s *SyncStore) ReadPuppy(id uint32) (*Puppy, error) {
+	if puppy, ok := s.syncStore.Load(id); ok {
+		returnPuppy := puppy.(Puppy)
+		return &returnPuppy, nil
+	}
+	return nil, &Error{
+		Message: "Puppy ID can not be found, read operation failed.",
+		Code:    2,
+	}
+}
+
+// UpdatePuppy updates the store with key of id with the new puppy.
+// It returns a boolean whether the operation is successful or not.
+func (s *SyncStore) UpdatePuppy(id uint32, puppy *Puppy) (bool, error) {
+	if _, ok := s.syncStore.Load(id); ok {
+		s.mu.Lock()
```

Comment:
My question here is whether or not the Lock should be done before the Load... in the example that @dan provided with lab6, it does the Lock as the first statement and then defer Unlock (as mentioned in my comment above)... I see ur logic in that u are only locking if u need to update but I think using the Lock/defer Unlock pattern is preferable.

Reply:
It might be a good idea to lock it before reading it, as I can think of a scenario where it can be deleted before you trying to update it. So I will lock it.

Reply:
hmmm, works fine for me... 
`// UpdatePuppy modifies puppy data in the sync store, either creating a new one or overwriting an old one.
func (s *SyncStore) UpdatePuppy(puppyID uint32, p *Puppy) (uint32, error) {
	s.Lock()
	defer s.Unlock()
	if _, ok := s.Load(puppyID); !ok {
		s.currID++
		puppyID = s.currID
	}
	p.ID = puppyID //ensure the ID within p always matches the sync store key (puppyID)
	s.Store(puppyID, *p)
	return puppyID, nil
}`
Do u want me to come and visit u?

Reply:
Sorry... I don't know how to insert code so that it displays properly

Reply:
@pfannerj Use triple backticks around the block of code instead of single backticks. I've edited your comment, so take a look

Reply:
@camh-anz  Thanks Cam, I told her the answer via slack yesterday.